### DISCUSSION TITLE
Waveshare esp32-s3 2.1 touch

### DISCUSSION THREAD

**Discussion by themagicm** [lovyan03/LovyanGFX#630] [Device Configurations]
Anyone have an example code that works with this device?

https://www.waveshare.com/esp32-s3-touch-lcd-2.1.htm

**Comment by tobozo** [comment]
the [wiki](https://www.waveshare.com/wiki/ESP32-S3-Touch-LCD-2.1) has a some [demo code](https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-2.1/ESP32-S3-Touch-LCD-2.1-Code.zip) both for esp-idf and arduino 

however there is no existing LGFX config for this device, so you'll have to build your own.

you can use [this configuration example](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx_user/LGFX_ESP32S3_RGB_MakerfabsParallelTFTwithTouch40.h) as a basis:

- keep Panel_ST7701 section
- replace `Touch_GT911` with [`Touch_CST816S`](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/touch/Touch_CST816S.cpp) 
- edit the pin numbers to match those from the wiki

**Comment by themagicm** [comment]
So the 816S is compatible with the 820 that is on this display?

Sent from Proton Mail Android

-------- Original Message --------

**Reply by tobozo** [reply]
yes

https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/touch/Touch_CST816S.cpp#L19-L20

**Comment by shark526** [comment]
hi @themagicm, have you make it work? I'm doing the same but still not work, confusing with EXIO_Init

**Comment by fallenartist** [comment]
Hi @themagicm I'd also like to know if you've had any success with this display as I'm trying desperately to draw anything on it, so far unsuccessfully. Ta!

**Reply by fallenartist** [reply]
@tobozo would you be so kind and point me in the right direction? So far I was able to get multicoloured jittering lines on the display ;( I followed the pin declarations from Waveshare example files but, as @shark526 mentioned, there's a "TCA9554PWR I2C GPIO expansion chip (EXIO)" used in demo and I'm not sure how to implement it with LGFX. BTW I reached similar result (sans the jitter) with Arduino GFX, adapting the demo code. Thanks in advance for any pointers.

Pin spec:
- LCD_BL | GPIO6
- LCD_RST | **EXIO1**
- LCD_SDA | GPIO1
- LCD_SCL | GPIO2
- LCD_CS | **EXIO3**
- PCLK | GPIO41
- DE | GPIO40
- VSYNC | GPIO39
- HSYNC | GPIO38
- B0 | NC
- B1 | GPIO5
- B2 | GPIO45
- B3 | GPIO48
- B4 | GPIO47
- B5 | GPIO21
- G0 | GPIO14
- G1 | GPIO13
- G2 | GPIO12
- G3 | GPIO11
- G4 | GPIO10
- G5 | GPIO9
- R0 | NC
- R1 | GPIO46
- R2 | GPIO3
- R3 | GPIO8
- R4 | GPIO18
- R5 | GPIO17
- TP_SDA | GPIO15
- TP_SCL | GPIO7
- TP_INT | GPIO16
- TP_RST | **EXIO2**

My minimal code:

```// LCD_Test.ino

#include <LovyanGFX.hpp>

#include "LGFX_ESP32S3_RGB_WaveshareLCD21.h" // ?

static LGFX lcd;

void setup(void)
{
  lcd.init();
  lcd.setRotation(1);
  lcd.setBrightness(128);
  lcd.setColorDepth(16);

  lcd.fillScreen(0xFF8844u);
}

void loop(void)
{
  // do nothing for now
}

// LGFX_ESP32S3_RGB_WaveshareLCD21.h:

#define LGFX_USE_V1
#include <LovyanGFX.hpp>

#include <lgfx/v1/platforms/esp32s3/Panel_RGB.hpp>
#include <lgfx/v1/platforms/esp32s3/Bus_RGB.hpp>
#include <driver/i2c.h>

class LGFX : public lgfx::LGFX_Device
{
public:

  lgfx::Bus_RGB      _bus_instance;
  lgfx::Panel_ST7701 _panel_instance;
  lgfx::Touch_CST816S  _touch_instance;
  lgfx::Light_PWM   _light_instance;

  LGFX(void)
  {
	{
	  auto cfg = _panel_instance.config();

	  cfg.memory_width  = 480;
	  cfg.memory_height = 480;
	  cfg.panel_width  = 480;
	  cfg.panel_height = 480;

	  cfg.offset_x = 0;
	  cfg.offset_y = 0;

	  _panel_instance.config(cfg);
	}

	{
	  auto cfg = _panel_instance.config_detail();

	  cfg.pin_cs = -1; // EXIO3 ?
	  cfg.pin_sclk = 2;
	  cfg.pin_mosi = 1;

	  _panel_instance.config_detail(cfg);
	}

	{
	  auto cfg = _bus_instance.config();
	  cfg.panel = &_panel_instance;
	  cfg.pin_d0  = GPIO_NUM_5;  // B0
	  cfg.pin_d1  = GPIO_NUM_45;  // B1
	  cfg.pin_d2  = GPIO_NUM_48; // B2
	  cfg.pin_d3  = GPIO_NUM_47; // B3
	  cfg.pin_d4  = GPIO_NUM_21;  // B4
	  cfg.pin_d5  = GPIO_NUM_14;  // G0
	  cfg.pin_d6  = GPIO_NUM_13;  // G1
	  cfg.pin_d7  = GPIO_NUM_12; // G2
	  cfg.pin_d8  = GPIO_NUM_11; // G3
	  cfg.pin_d9  = GPIO_NUM_10; // G4
	  cfg.pin_d10 = GPIO_NUM_9;  // G5
	  cfg.pin_d11 = GPIO_NUM_46; // R0
	  cfg.pin_d12 = GPIO_NUM_3; // R1
	  cfg.pin_d13 = GPIO_NUM_8; // R2
	  cfg.pin_d14 = GPIO_NUM_18; // R3
	  cfg.pin_d15 = GPIO_NUM_17; // R4

	  cfg.pin_henable = GPIO_NUM_40; // DE?
	  cfg.pin_vsync   = GPIO_NUM_39;
	  cfg.pin_hsync   = GPIO_NUM_38;
	  cfg.pin_pclk    = GPIO_NUM_41;
	  cfg.freq_write  = 14000000;

	  cfg.hsync_polarity    = 0;
	  cfg.hsync_front_porch = 50;
	  cfg.hsync_pulse_width = 8;
	  cfg.hsync_back_porch  = 8;
	  cfg.vsync_polarity    = 0;
	  cfg.vsync_front_porch = 8;
	  cfg.vsync_pulse_width = 3;
	  cfg.vsync_back_porch  = 8;
	  cfg.pclk_idle_high    = 0;
	  cfg.de_idle_high      = 1;
	  _bus_instance.config(cfg);
	}
	_panel_instance.setBus(&_bus_instance);

	{
	  auto cfg = _touch_instance.config();
	  cfg.x_min      = 0;
	  cfg.x_max      = 480;
	  cfg.y_min      = 0;
	  cfg.y_max      = 480;
	  cfg.bus_shared = false;
	  cfg.offset_rotation = 0;

	  cfg.i2c_port   = I2C_NUM_1;

	  cfg.pin_int    = GPIO_NUM_NC;
	  cfg.pin_sda    = GPIO_NUM_1;
	  cfg.pin_scl    = GPIO_NUM_2;
	  cfg.pin_rst    = GPIO_NUM_NC; // EXIO1 ?

	  cfg.freq       = 400000;
	  _touch_instance.config(cfg);
	  _panel_instance.setTouch(&_touch_instance);
	}

	{
	  auto cfg = _light_instance.config();
	  cfg.pin_bl = GPIO_NUM_6;
	  _light_instance.config(cfg);
	}
	_panel_instance.light(&_light_instance);

	setPanel(&_panel_instance);
  }
};```

**Comment by tobozo** [comment]
warning: your code sample includes LovyanGFX twice, make sure both the include and the LGFX config block are placed before setup() and loop(), and only include <LovyanGFX.hpp> once

According to Waveshare zip example, this init sequence handles the backlight, it is called before initializing the LCD:

```cpp

  I2C_Init();
  TCA9554PWR_Init(0x00);   
  Set_EXIO(EXIO_PIN8,Low);

```

this is the equivalent of doing this in your setup before tft.init():

```cpp
  Wire.begin(sda, scl); 
  Wire.beginTransmission(0x20); // I2C device ID
  Wire.write(0x03); // EXIO config reg
  Wire.write(0x00); // EXIO config value
  I2C.endTransmission();
  digitalWrite(EXIO_PIN8, LOW);
```

jittering: most parallel display issues can be solved by tweaking one of those settings:

```Cpp
	  cfg.hsync_polarity    = 0;
	  cfg.hsync_front_porch = 50;
	  cfg.hsync_pulse_width = 8;
	  cfg.hsync_back_porch  = 8;
	  cfg.vsync_polarity    = 0;
	  cfg.vsync_front_porch = 8;
	  cfg.vsync_pulse_width = 3;
	  cfg.vsync_back_porch  = 8;
	  cfg.pclk_idle_high    = 0;
	  cfg.de_idle_high      = 1;
```

**Reply by fallenartist** [reply]
Thanks for a quick reply. Sorry, there are two files joined together in my code block. There's a separate include from `// LGFX_ESP32S3_RGB_WaveshareLCD21.h:` line.

Can you expand on the EXIO, please? I'm not sure what I should do to actually use it in config for LCD_RST (EXIO1), LCD_CS (EXIO3) and TP_RST (EXIO2, touch?). Do I have to write a function? It's a bit advanced for me I'm afraid. 

The backlight is on pin 6, I believe, hence `cfg.pin_bl = GPIO_NUM_6;`?

Thank you.

**Reply by tobozo** [reply]
:warning:  `cfg_pin_bl` is for the backlight signal, not for powering it. ( a GPIO pin couldn't possibly provide that to a LCD panel)

LovyanGFX can drive the backlight signal, but it's up to you to handle whatever power module comes with your board, and have it enabled before using LGFX::init().

Also I don't know this EXIO module at all, my comments were exclusively inspired by what I read in the [demo code](https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-2.1/ESP32-S3-Touch-LCD-2.1-Code.zip) provided by Waveshare, which I linked earlier.

**Reply by fallenartist** [reply]
@tobozo sure, I'm just trying to understand it all and need a bit of guidance as I'm way out of my depth here. I hoped for the board to just work and focus on the front-end (designer here). Anyway, from the demo code I gathered the LCD init is as follows:
```
  I2C_Init(); // Wire.begin(15, 7); 
  TCA9554PWR_Init(0x00); // Wire
  Set_EXIO(EXIO_PIN8, Low); // EXIO8 is buzzer
  ST7701_Reset(); // EXIO1 off and on?
  {
    Set_EXIO(EXIO_PIN1, Low);
    vTaskDelay(pdMS_TO_TICKS(10));
    Set_EXIO(EXIO_PIN1, High);
    vTaskDelay(pdMS_TO_TICKS(50));
  }
  ST7701_Init();
  {
    ST7701_CS_EN(); // EXIO3 off
    {
      Set_EXIO(EXIO_PIN3, Low);
      vTaskDelay(pdMS_TO_TICKS(10));
    }
    // 
    // big init command sequence here
    // 
    ST7701_CS_Dis(); // EXIO3 back on?
    {
      Set_EXIO(EXIO_PIN3, High);
      vTaskDelay(pdMS_TO_TICKS(10));
    }
  } 
  Touch_Init(); // I2C, CST820_ADDR 0x15, CST820_INT_PIN 16
  Backlight_Init(); // ledcAttach(6, 20000, 10); ledcWrite(6, Duty);
}
```

My question is do you see any potential conflicts with LovyanGFX library?
The screen now got black, with backlight on and touch responding fine. 
Is it all just finding the "holy grail" of the exact initialisation sequence or is there more to it?
Thanks again for your help with this.

**Reply by tobozo** [reply]
potential conflicts may occur with shared bus, but since you've successfully had touch and backlight under simultaneous control I guess it's working fine.

> The screen now got black

is that a regression or is it progress? you mentioned jittering earlier, what did it look like?

**Reply by fallenartist** [reply]
I think both. I surely had the pins mixed up when the jittering stripes showed but I can't replicate that again. So now, with the correct pins and porch values (as in demo) the screen is backlit, black and registers touch. The stripes were just horizontal (or vertical?) 1px lines, different random colour of each line.

I actually asked AI to update the original demo and strip it from LVGL and several iterations later it works (using IDF/esp_lcd driver (`esp_lcd_new_rgb_panel()`).

Would you be able to [look at the code](https://gist.github.com/fallenartist/22d1d01e125afb02ae4ebdcdf02d1f80) and let me know if this can be translated to LGFX?

**Reply by sandmen** [reply]
J> I think both. I surely had the pins mixed up when the jittering stripes showed but I can't replicate that again. So now, with the correct pins and porch values (as in demo) the screen is backlit, black and registers touch. The stripes were just horizontal (or vertical?) 1px lines, different random colour of each line.
> 
> I actually asked AI to update the original demo and strip it from LVGL and several iterations later it works (using IDF/esp_lcd driver (`esp_lcd_new_rgb_panel()`).
> 
> Would you be able to [look at the code](https://gist.github.com/fallenartist/22d1d01e125afb02ae4ebdcdf02d1f80) and let me know if this can be translated to LGFX?

Did your Code on gist working with the Display and lovyangfx ?

Or do you have a working Setup with this Display and lovyangfx ?

**Comment by palis-ss** [comment]
The purpose of the CS pin is that it is used during SPI transactions between the MCU and the ST7701.  Normally, we would connect the CS pin directly to a GPIO of the MCU, which LGFX wants us to put that pin no in the configuration struct during initialization, e.g., `cfg.pin_cs = xxx;`.  Unfortunately, for this particular board, the CS pin is connected to EXIO3 which is an output from the IO Exander chip (TCA9554).  This chip is controlled via I2C, i.e., we send a command via I2C to set/unset a pin.  That is when you see calls to the function `Set_EXIO()` in the demo code.   Since it's not a direct control via GPIO, we have to find some other way so LGFX knows how to set/unset the CS pin differently.

BTW, the CS pin will get set/unset a lot, not just during initialization.  Putting in just the init function won't do it.
